### PR TITLE
Bump rust-toolchain to `nightly-2021-07-28`

### PR DESCRIPTION
**Pull request by Aaron1011**
### What problem does this PR solve?

This will hopefully fix the incremental compilation issue
described in https://github.com/rust-lang/rust/issues/87516

Problem Summary:

Bump `rust-toolchain` to the latest nighty (as of the time this PR was opened): `nightly-2021-07-28`

What's Changed:
* Some feature flags were adjusted
* The `assert_matches` is no longer in the prelude
* `as_ne_bytes` was removed

### Check List 

Side effects

- tikv can no longer be built with `nightly-2021-04-15`

### Release note 

```release-note
Update required Rust version to `nightly-2021-07-28`
```

### CODE REVIEW

**Comment by Ray-Eldath** [comment]
this effectively fix https://github.com/rust-lang/rust/issues/87516.

**Comment by hicqu** [comment]
There are still some clippy warnnings for `FAIL_POINT=1 make clippy`:
```
warning: variant name ends with the enum's name
  --> src/server/service/diagnostics/log.rs:38:5
   |
38 |     ParseError(String),
   |     ^^^^^^^^^^^^^^^^^^
   |
   = note: `#[warn(clippy::enum_variant_names)]` on by default
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#enum_variant_names

warning: variant name ends with the enum's name
  --> src/server/service/diagnostics/log.rs:39:5
   |
39 |     SearchError(String),
   |     ^^^^^^^^^^^^^^^^^^^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#enum_variant_names

warning: variant name ends with the enum's name
  --> src/server/service/diagnostics/log.rs:40:5
   |
40 |     IOError(std::io::Error),
   |     ^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#enum_variant_names
```

Would you mind to fix them? @Aaron1011

**Comment by hicqu** [comment]
And please also fix this:
```
$ git diff
diff --git a/tests/failpoints/cases/test_gc_worker.rs b/tests/failpoints/cases/test_gc_worker.rs
index c63ef7061..4b3dbf59b 100644
--- a/tests/failpoints/cases/test_gc_worker.rs
+++ b/tests/failpoints/cases/test_gc_worker.rs
@@ -249,7 +249,7 @@ fn test_collect_applying_locks() {

     // Write 1 lock.
     must_kv_prewrite(
-        &store_1_client,
+        store_1_client,

```

**Comment by Aaron1011** [comment]
@hicqu: Updated

**Comment by sticnarf** [comment]
@Aaron1011 I find there is still an open PR fixing incremental compilation by you: https://github.com/rust-lang/rust/pull/85868.
Do you think it a good idea to wait until that PR being merged?

**Comment by Aaron1011** [comment]
@sticnarf That PR fixes a pre-existing bug in incremental compilation, which affects the version of Rust that you're currently using. The version used by this PR has several other incremental compilation fixes included, so it's still worthwhile to upgrade even without the additional fix.

Also, I'm not sure when that PR you linked will be merged.

**Review by sticnarf — COMMENTED** [review]
Please resolve the conflict. The rest LGTM.

A lot thanks for fixing incremental compilation.

**Review by hicqu — APPROVED** [review]

**Comment by Aaron1011** [comment]
@sticnarf: Rebasd

**Review by sticnarf — APPROVED** [review]

File: components/cdc/src/endpoint.rs
Comment (skyzh) [inline components/cdc/src/endpoint.rs:907]: Just curious, why could we remove the `&` here?

Comment (sticnarf) [reply components/cdc/src/endpoint.rs:907]: Sorry but why do you think we need the `&`?

Comment (skyzh) [reply components/cdc/src/endpoint.rs:907]: Thanks for explanation. So the `&` is redundant even before this change.

File: scripts/clippy
Comment (BusyJay) [inline scripts/clippy:38]: I believe this `--fix` breaks CI checks. In CI, it relies on non-zero exit code to check if code are formatted and linted. Now it always passes even the code breaks.

**Pull request merged**
